---
title: "Extreme phenotype sampling"
author: "Thea Bjornland"
date: "30 June 17"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Extreme phenotype sampling}
  %\VignetteEncoding{UTF-8}
---

This package provides functions for fitting linear regression models
and testing hypothesis of genetic association with trait
for data from the extreme phenotype sampling (EPS) design.

For details on the EPS design and methods, please see our paper "Improving power of genetic association studies by extreme phenotype sampling: a review and some new results"; 
arXiv preprint arXiv:1701.01286 (https://arxiv.org/abs/1701.01286)


```{r}
    library(extremesampling)
```

The assumed linear regression model is of the form
`y = a + be*xe + bg*xg + e`, where `e` is normal distributed with mean 0 and 
standard deviation `sigma`.

Here, `xg` represents genetic covariates in the form of SNPs (single-nucleotide
polymorphisms), while `xe` represents non-genetic covariates, if any. 
Additionally, gene-environment interaction terms `xe*xg` can be included. 

In EPS, selective genotyping is based on extreme phenotype
individuals. Extreme phenotypes are defined by lower and upper cutoffs, `l` 
and `u`. SNP genotypes are obtained for individuals with `y < l` 
or `y > u`, and possibly a random sample.

This package handles two types of extreme sampling data: 
**EPS complete-case** or **EPS all-case**.

### EPS complete-case (EPS-CC)

For the **EPS complete-case** design, variables (`y,xe,xg`) are available only 
for extreme phenotype individuals (and possibly a random sample). 
Furthermore, the values of the lower and upper cutoffs (`l,u`) must be known.

The following data set will be used for illustrating EPS-CC functions. 

```{r}
    # Generate data set of the type EPS-CC

    N = 5000 # Number of individuals in a population
    xe1 = rnorm(n = N, mean = 2, sd = 1) # Environmental covariate
    xe2 = rbinom(n = N, size = 1, prob = 0.3) # Environmental covariate
    xg1 = sample(c(0,1,2),N,c(0.4,0.3,0.3), replace = TRUE) # SNP
    xg2 = sample(c(0,1,2),N,c(0.5,0.3,0.2), replace = TRUE) # SNP
    
    # Model parameters
    a = 50; be1 = 5; be2 = 8; bg1 = 0.3; bg2 = 0.6; sigma = 2
    
    # Generate response y (phenotype)
    y = rnorm(N, mean = a+be1*xe1+be2*xe2+bg1*xg1+bg2*xg2, sd = sigma)
    
    # Identify extremes, here upper and lower 25% of population
    u = quantile(y,probs = 3/4,na.rm=TRUE)
    l = quantile(y,probs = 1/4,na.rm=TRUE)
    extreme = (y < l) | (y >= u)
    
    # Create the EPS-only data set
    y = y[extreme]
    xe1 = xe1[extreme]
    xe2 = xe2[extreme]
    xg1 = xg1[extreme]
    xg2 = xg2[extreme]
    
    # Matrix form of covariates
    xg = as.matrix(cbind(xg1,xg2))
    xe = as.matrix(cbind(xe1,xe2))
```

First we consider testing in the EPS-CC design. Here are a few key examples.

Example 1.1) Test `H0: bg = 0` against `H1: bg ≠ 0` for 
each genetic covariate at a time. Each genetic variable is 
tested against the null model `y = a + be*xe + e`.

```{r}
    # 1.1) Perform two-sided test for one genetic covariate at a time
    epsCC.test(y~xe1+xe2, xg=cbind(xg1,xg2), cutoffs=c(l,u))$p.value
    # or using matrix form
    # epsCC.test(y~xe, xg=xg, cutoffs=c(l,u))$p.value
```

Example 1.2) Test gene-environment interactions

```{r}
    # 1.2) Test gene-environment interaction
    epsCC.test(y~xe1+xe2+xg1+xg2, xg = xe1*xg1, cutoffs=c(l,u))$p.value
```

Next we consider fitting a linear regression model of the form
`y = a + be*xe + bg*xg + e` using data from the EPS-CC design. 

Example 1.3) Fit a linear regression model to obtain parameter estimates 
and 95% confidence intervals

```{r}
    # 1.3) Fit a linear model
    epsCC.lm(y~xe1+xe2+xg1+xg2, cutoffs=c(l,u))
    # alternatively with matrix inputs
    # epsCC.lm(y~xe+xg, cutoffs=c(l,u))
```

Example 1.4) Fit a linear regression model with an interaction term

```{r}
    # 1.4) Fit a linear model with an interaction term
    epsCC.lm(y~xe1+xe2+xg1+xg2+xe2*xg1, cutoffs=c(l,u))$coefficients
```

### EPS all-case (EPS-AC)

For the **EPS all-case design**, the variables `y` and `xe` are observed for 
all individuals (in a population / random sample) while `xg` is observed only
for the extreme phenotype individuals. The value of `xg` should be set to 
`NA` for non-extreme phenotype individuals for which the genotype is not
observed. The functions for testing and
model fitting are based on a likelihood model for missing at random (MAR) 
covariates when the missing-mechanism is ignorable for likelihood inference. 

The following example will be used throughout for illustrating EPS-AC
functions. 

```{r}
    # Generate example data set of the type EPS-AC

    N = 5000 # Number of individuals in a population
    xe1 = rnorm(n = N, mean = 2, sd = 1) # Environmental covariate
    xe2 = rbinom(n = N, size = 1, prob = 0.3) # Environmental covariate
    xg1 = sample(c(0,1,2), N, c(0.4,0.3,0.3), replace=TRUE) # SNP
    xg2 = sample(c(0,1,2), N, c(0.5,0.3,0.2), replace=TRUE) # SNP
    
    # Model parameters
    a = 50; be1 = 5; be2 = 8; bg1 = 0.3; bg2 = 0.6; sigma = 2
    
    # Generate response y
    y = rnorm(N, mean = a + be1*xe1 + be2*xe2 + bg1*xg1 + bg2*xg2, sd = sigma)
    
    # Identify extremes, here upper and lower 25% of population
    u = quantile(y,probs = 3/4,na.rm=TRUE)
    l = quantile(y,probs = 1/4,na.rm=TRUE)
    extreme = (y < l) | (y >= u)
    
    # Create the EPS-full data set by setting
    # the SNP values of non-extremes to NA
    xg1[!extreme] = NA
    xg2[!extreme] = NA
    
    xg = as.matrix(cbind(xg1,xg2))
    xe = as.matrix(cbind(xe1,xe2))
```


Below, we present a few key examples of testing for the EPS-AC design.

Example 2.1) Test `H0: bg = 0` against `H1: bg ≠ 0`. Each genetic variable is 
considered for inclusion into the  null model `y = a + be*xe + e`.

```{r}
    # 2.1) Perform score test for one genetic covariate at a time
    epsAC.test(y~xe1+xe2, xg=cbind(xg1,xg2))$p.value
    # alternatively with matrix form
    epsAC.test(y~xe, xg=xg)$p.value
```

Example 2.2) Test a gene-environment interaction
```{r}
    # 2.2) Test interactions
    epsAC.test.GE(y~xe1+xe2+xg1, G = xg1, E = xe1)$p.value
```

Next we consider fitting the linear regression model 
`y = a + be*xe + bg*xg + e` using data from the EPS-AC design.

Example 2.3) Fit a linear regression model to obtain parameter estimates 
and 95% confidence intervals
```{r}
    # 2.4) Fit a linear model
    epsAC.lm(y~xe1+xe2+xg1+xg2)
    # alternatively with matrix inputs
    # epsAC.lm(y~xe+xg)
```

Example 2.5) Fit a linear regression model with gene-environment interactions

```{r}
    # 2.5) Fit a linear model with an interaction term
    epsAC.lm(y~xe1+xe2+xg1+xg2+xe1*xg2)$coefficients
```

#### EPS-AC and confounding effects

If there are confounding effects in the data, then this must
be accounted for in the distribution of `xg` in the EPS-full 
model. For example, if the distribution of a SNP in the population is 
different between men and women, then the estimation of the distribution of 
the SNP genotype must be conditioned on the sex of each individual. 
If not, the SNP can become a mediator for the effect
of gender on phenotype, which will result in bias and false positives. 

The functions for EPS-AC with confounding lets the user type in which 
environmental covariates `xe` that are confounders for SNPs. In this package,
confounders can only be discrete environmental covariates.

The following data set is created such that one SNP is dependent upon a 
binary covariate. The tests for associations show that a spurious results
are found when confounding is not accounted for.


```{r}
    # Generate example data set of the type EPS-AC
    # with confounding

    N = 5000 # Number of individuals in a population
    xe1 = rnorm(n = N, mean = 2, sd = 1) # Environmental covariate
    xe2 = rbinom(n = N, size = 1, prob = 0.3) # Environmental covariate
    xg1 = sample(c(0,1,2), N, c(0.4,0.3,0.3), replace = TRUE) # SNP
    
    xg2 = rep(NA,N) # SNP
    n0 = sum(xe2 == 0)
    n1 = sum(xe2 == 1)
    xg2[xe2 == 0] = sample(c(0,1,2),n0,c(0.5,0.3,0.2), replace = TRUE)
    xg2[xe2 == 1] = sample(c(0,1,2),n1,c(0.3,0.1,0.6), replace = TRUE)
    
    # Model parameters
    a = 50; be1 = 5; be2 = 8; bg1 = 0.3; sigma = 2
    
    # Generate response y (xg2 has no effect)
    y = rnorm(N, mean = a + be1*xe1 + be2*xe2 + bg1*xg1 , sd = sigma)
    
    # Identify extremes, here upper and lower 25% of population
    u = quantile(y,probs = 3/4,na.rm=TRUE)
    l = quantile(y,probs = 1/4,na.rm=TRUE)
    extreme = (y < l) | (y >= u)
    
    xg1[!extreme] = NA
    xg2[!extreme] = NA
    
    xg = cbind(xg1,xg2)
    
    # Test (with confounding)
    epsAC.test(y~xe1+xe2, xg=xg, confounder=c("xe2"))
    
    # Test (without confounding)
    epsAC.test(y~xe1+xe2, xg=xg)
```

