---
title: "Extreme phenotype sampling"
author: "Thea Bjornland"
date: "30 June 17"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Extreme phenotype sampling}
  %\VignetteEncoding{UTF-8}
---

This package provides functions for testing hypothesis of genetic association with trait
for data from the extreme phenotype sampling (EPS) design.

For details on the EPS design and methods, please see our paper "Improving power of genetic association studies by extreme phenotype sampling: a review and some new results"; 
arXiv preprint arXiv:1701.01286 (https://arxiv.org/abs/1701.01286)


```{r}
    library(extremesampling)
```

The assumed linear regression model is of the form
`y = a + be*xe + bg*xg + e`, where `e` is normal distributed with mean 0 and 
standard deviation `sigma`.

Here, `xg` represents genetic covariates in the form of SNPs (single-nucleotide
polymorphisms), while `xe` represents non-genetic covariates, if any. 

In EPS, selective genotyping is based on extreme phenotype
individuals.

# Y-extreme sampling
Extreme phenotypes are defined by lower and upper cutoffs, `l` 
and `u` and SNP genotypes are obtained for individuals with `y < l` 
or `y > u`.

### EPS complete-case (EPS-CC)

For the **EPS complete-case** method, variables (`y,xe,xg`) are only 
considered for for extreme phenotype individual.  The values of the lower and 
upper cutoffs (`l,u`) must be known.

The following data will be used for illustrating EPS-CC functions. 

```{r}
    # Generate data set of the type EPS-CC

    N = 5000 # Number of individuals in a population
    xe1 = rnorm(n = N, mean = 2, sd = 1) # Environmental covariate
    xe2 = rbinom(n = N, size = 1, prob = 0.3) # Environmental covariate
    xg1 = rbinom(N,2,0.4) # SNP
    xg2 = rbinom(N,2,0.1) # SNP
    
    # Model parameters
    a = 50; be1 = 5; be2 = 8; bg1 = 0.3; bg2 = 0.6; sigma = 2
    
    # Generate response y (phenotype)
    y = rnorm(N, mean = a+be1*xe1+be2*xe2+bg1*xg1+bg2*xg2, sd = sigma)
    
    # Identify extremes, here upper and lower 25% of population
    u = quantile(y,probs = 3/4,na.rm=TRUE)
    l = quantile(y,probs = 1/4,na.rm=TRUE)
    extreme = (y < l) | (y >= u)
    
    # Create the EPS-only data set
    y = y[extreme]
    xe1 = xe1[extreme]
    xe2 = xe2[extreme]
    xg1 = xg1[extreme]
    xg2 = xg2[extreme]
    
    # Matrix form of covariates
    xg = as.matrix(cbind(xg1,xg2))
    xe = as.matrix(cbind(xe1,xe2))
```

First we consider testing in the EPS-CC design. Here are a few key examples.

Example 1.1) Test $H0: \beta_g = 0$ against $H1: \beta_g \neq 0$ for 
each genetic covariate at a time. Each genetic variable is 
tested against the null model `y = a + be*xe + e`.

```{r}
    # 1.1) Perform two-sided test for one genetic covariate at a time
    epsCC.test(y~xe1+xe2, xg=cbind(xg1,xg2), l,u)$p.value
    # or using matrix form
    # epsCC.test(y~xe, xg=xg, cutoffs=c(l,u))$p.value
```

