---
title: "Extreme phenotype sampling"
author: "Thea Bjornland"
date: "30 June 17"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{Extreme phenotype sampling}
  %\VignetteEncoding{UTF-8}
---

This package provides functions for testing hypothesis of genetic association with a continously measurable trait for data from the extreme phenotype sampling (EPS) design.


```{r}
    library(extremesampling)
```

The assumed linear regression model is of the form
`y = a + be*xe + bg*xg + e`, where `e` is normal distributed with mean 0 and 
standard deviation `sigma`.

Here, `xg` represents genetic covariates in the form of SNPs (single-nucleotide
polymorphisms), while `xe` represents non-genetic covariates, if any. 

Extreme phenotypes are defined by lower and upper cutoffs, `l` 
and `u` and SNP genotypes are obtained only for individuals with `y < l` 
or `y > u`. The cut-offs can be specific for each individual.

## Single-variant association tests

The following data set is meant to illustrate single-variant association testing under the
z-extreme sampling design.

```{r}
    N = 5000 # Number of individuals in a population
    xe1 = rnorm(n = N, mean = 2, sd = 1) # Environmental covariate
    xe2 = rbinom(n = N, size = 1, prob = 0.3) # Environmental covariate
    xg1 = rbinom(N,2,0.25) # SNP
    
    # Generate phenotype
    y = rnorm(N, mean = 10+5*xe1+1*xe2+0.5*xg1, sd = 4)
    
    # Identify extremes, here upper and lower 25% of residual phenotype distribution
    z = lm(y~xe1+xe2)$residuals
    uz = quantile(z,probs = 3/4,na.rm=TRUE)
    lz = quantile(z,probs = 1/4,na.rm=TRUE)
    extreme = (z < lz) | (z >= uz)
    li = lz + (y[extreme]-z[extreme]); ui = uz + (y[extreme]-z[extreme])
    
    # Create the EPS complete case data set
    yCC = y[extreme]; zCC = z[extreme]
    xe1CC = xe1[extreme]; xe2CC = xe2[extreme]
    xg1CC = xg1[extreme]
    
    # Create the EPS all case data set
    yAC = y; zAC = z
    xe1AC = xe1; xe2AC = xe2
    xg1AC = xg1; xg1AC[!extreme] = NA
```

### Complete case tests

```{r}

cc = epsCC.test(yCC~xe1CC + xe2CC, xg = xg1CC, l = li, u = ui)
cc$statistic
cc$p.value

# The complete case test under z-extreme sampling is (at least asymptotically) 
# equal to the score test for random samples. 
# We show this by using glm.scoretest in the Statmod package.

library(statmod)
lm0 = glm(yCC~xe1CC + xe2CC)
tRand = glm.scoretest(lm0,xg1CC)^2
pRand = pchisq(tRand,1,lower.tail = FALSE)
tRand
pRand

# Residual-based test (equal to above test when xg is not correlated with xe)

ccres= epsCC.test(zCC~1, xg = xg1CC, l = lz, u = uz)
ccres$statistic
ccres$p.value

```

### All case tests

```{r}
# The all case test is approximately equal to the complete case test under z-extreme sampling
ac = epsAC.test(yAC~xe1AC + xe2AC, xg = xg1AC)
ac$statistic
ac$p.value

# If confounding assumed (xe2 discrete non-genetic covariate)
ac_conf = epsAC.test(yAC~xe1AC + xe2AC, xg = xg1AC, confounder = xe2AC)
ac_conf$statistic
ac_conf$p.value

# Residual-based test 
acres = epsAC.test(zAC~1, xg = xg1AC)
acres$statistic
acres$p.value

```

## Multiple-variant association tests
These tests are especially derived for rare variant association testing. 

```{r}
    N = 5000 # Number of individuals in a population
    xe1 = rnorm(n = N, mean = 2, sd = 1) # Environmental covariate
    xe2 = rbinom(n = N, size = 1, prob = 0.3) # Environmental covariate
    Gmat = cbind(rbinom(N,2,0.04),rbinom(N,2,0.03),rbinom(N,2,0.02),
                 rbinom(N,2,0.01),rbinom(N,2,0.01),rbinom(N,2,0.01)) # rare variants
    
    # Generate phenotype
    betag = c(0.1,-0.1,0.3,0.5,-0.2,0)
    y = rnorm(N, mean = 10+5*xe1+1*xe2+Gmat%*%betag, sd = 4)
    
    # Identify extremes, here upper and lower 25% of residual phenotype distribution
    z = lm(y~xe1+xe2)$residuals
    uz = quantile(z,probs = 3/4,na.rm=TRUE)
    lz = quantile(z,probs = 1/4,na.rm=TRUE)
    extreme = (z < lz) | (z >= uz)
    li = lz + (y[extreme]-z[extreme]); ui = uz + (y[extreme]-z[extreme])
    
    # Create the EPS complete case data set
    yCC = y[extreme]; zCC = z[extreme]
    xe1CC = xe1[extreme]; xe2CC = xe2[extreme]
    GmatCC = Gmat[extreme,]
    
    # Create the EPS all case data set
    yAC = y; zAC = z
    xe1AC = xe1; xe2AC = xe2
    GmatAC = Gmat; GmatAC[!extreme,] = NA
```

### Complete case tests

```{r}

# Direct test of all 6 variants
epsCC.rv.test(yCC~xe1CC + xe2CC, xg = GmatCC, l = li, u = ui)$p.value

# Collapsing test (weights = 1 for all variants)
epsCC.rv.test(yCC~xe1CC + xe2CC, xg = GmatCC, l = li, u = ui, method = "collapse")$p.value

# Variance component test (weights = 1 for all variants)
epsCC.rv.test(yCC~xe1CC + xe2CC, xg = GmatCC, l = li, u = ui, method = "varcomp")$p.value


# Residual-based tests
# Direct test of all 6 variants
epsCC.rv.test(zCC~1, xg = GmatCC, l = lz, u = uz)$p.value

# Collapsing test (weights = 1 for all variants)
epsCC.rv.test(zCC~1, xg = GmatCC, l = lz, u = uz, method = "collapse")$p.value

# Variance component test (weights = 1 for all variants)
epsCC.rv.test(zCC~1, xg = GmatCC, l = lz, u = uz, method = "varcomp")$p.value


```

### All case tests

```{r}

# Direct test of all 6 variants
epsAC.rv.test(yAC~xe1AC + xe2AC, xg = GmatAC)$p.value

# Collapsing test (weights = 1 for all variants)
epsAC.rv.test(yAC~xe1AC + xe2AC, xg = GmatAC, method = "collapse")$p.value

# Variance component test (weights = 1 for all variants)
epsAC.rv.test(yAC~xe1AC + xe2AC, xg = GmatAC, method = "varcomp")$p.value

# Residual-based tests
# Direct test of all 6 variants
epsAC.rv.test(zAC~1, xg = GmatAC)$p.value

# Collapsing test (weights = 1 for all variants)
epsAC.rv.test(zAC~1, xg = GmatAC, method = "collapse")$p.value

# Variance component test (weights = 1 for all variants)
epsAC.rv.test(zAC~1, xg = GmatAC, method = "varcomp")$p.value

```

