<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Thea Bjornland" />


<title>Extreme phenotype sampling</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Extreme phenotype sampling</h1>
<h4 class="author"><em>Thea Bjornland</em></h4>
<h4 class="date"><em>3 Jan 17</em></h4>



<p>This package provides functions for fitting linear regression models and testing hypothesis of genetic association with trait for data from the extreme phenotype sampling (EPS) design.</p>
<p>For details on the EPS design and methods, please see our paper “Improving power of genetic association studies by extreme phenotype sampling: a review and some new results”; arXiv preprint arXiv:1701.01286 (<a href="https://arxiv.org/abs/1701.01286" class="uri">https://arxiv.org/abs/1701.01286</a>)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="kw">library</span>(extremesampling)</code></pre></div>
<p>The assumed linear regression model is of the form <code>y = a + be*xe + bg*xg + e</code>, where <code>e</code> is normal distributed with mean 0 and standard deviation <code>sigma</code>.</p>
<p>Here, <code>xg</code> represents genetic covariates in the form of SNPs (single-nucleotide polymorphisms), while <code>xe</code> represents non-genetic covariates, if any. Additionally, gene-environment interaction terms <code>xe*xg</code> can be included.</p>
<p>In EPS, selective genotyping is based on extreme phenotype individuals. Extreme phenotypes are defined by lower and upper cutoffs, <code>l</code> and <code>u</code>. SNP genotypes are only obtained for individuals with <code>y &lt; l</code> or <code>y &gt; u</code>.</p>
<p>This package provides functions for genetic association studies. For a genomewide association study, one SNP should be tested at a time against a null model. For a candidate SNP study with few SNPs, all SNPs can be tested simultaneously and the parameters of the regression model can be estimated. Furthermore, the functions allow for testing of gene-environment interactions, and estimate parameters in a gene-environment interaction model.</p>
<p>The functions in this package rely on the assumption of additive genetic models. Therefore, SNPs should take numerical values 0, 1 and 2 according to the number of minor-alleles. The effect of having one minor-allele is then <code>1*bg</code> while the effect of two minor-alleles is <code>2*bg</code>.</p>
<p>This package handles two types of extreme sampling data: <strong>EPS-only</strong> or <strong>EPS-full</strong>.</p>
<div id="eps-only" class="section level3">
<h3>EPS-only</h3>
<p>For the <strong>EPS-only</strong> design, all variables (<code>y,xe,xg</code>) are available only for extreme phenotype individuals. Furthermore, the values of the lower and upper cutoffs (<code>l,u</code>) must be known.</p>
<p>The following data set will be used for illustrating EPS-only functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># Generate data set of the type EPS-only</span>

    N =<span class="st"> </span><span class="dv">5000</span> <span class="co"># Number of individuals in a population</span>
    xe1 =<span class="st"> </span><span class="kw">rnorm</span>(<span class="dt">n =</span> N, <span class="dt">mean =</span> <span class="dv">2</span>, <span class="dt">sd =</span> <span class="dv">1</span>) <span class="co"># Environmental covariate</span>
    xe2 =<span class="st"> </span><span class="kw">rbinom</span>(<span class="dt">n =</span> N, <span class="dt">size =</span> <span class="dv">1</span>, <span class="dt">prob =</span> <span class="fl">0.3</span>) <span class="co"># Environmental covariate</span>
    xg1 =<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>),N,<span class="kw">c</span>(<span class="fl">0.4</span>,<span class="fl">0.3</span>,<span class="fl">0.3</span>), <span class="dt">replace =</span> <span class="ot">TRUE</span>) <span class="co"># SNP</span>
    xg2 =<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>),N,<span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">0.3</span>,<span class="fl">0.2</span>), <span class="dt">replace =</span> <span class="ot">TRUE</span>) <span class="co"># SNP</span>
    
    <span class="co"># Model parameters</span>
    a =<span class="st"> </span><span class="dv">50</span>; be1 =<span class="st"> </span><span class="dv">5</span>; be2 =<span class="st"> </span><span class="dv">8</span>; bg1 =<span class="st"> </span><span class="fl">0.3</span>; bg2 =<span class="st"> </span><span class="fl">0.6</span>; sigma =<span class="st"> </span><span class="dv">2</span>
    
    <span class="co"># Generate response y (phenotype)</span>
    y =<span class="st"> </span><span class="kw">rnorm</span>(N, <span class="dt">mean =</span> a+be1*xe1+be2*xe2+bg1*xg1+bg2*xg2, <span class="dt">sd =</span> sigma)
    
    <span class="co"># Identify extremes, here upper and lower 25% of population</span>
    u =<span class="st"> </span><span class="kw">quantile</span>(y,<span class="dt">probs =</span> <span class="dv">3</span>/<span class="dv">4</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
    l =<span class="st"> </span><span class="kw">quantile</span>(y,<span class="dt">probs =</span> <span class="dv">1</span>/<span class="dv">4</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
    extreme =<span class="st"> </span>(y &lt;<span class="st"> </span>l) |<span class="st"> </span>(y &gt;=<span class="st"> </span>u)
    
    <span class="co"># Create the EPS-only data set</span>
    y =<span class="st"> </span>y[extreme]
    xe1 =<span class="st"> </span>xe1[extreme]
    xe2 =<span class="st"> </span>xe2[extreme]
    xg1 =<span class="st"> </span>xg1[extreme]
    xg2 =<span class="st"> </span>xg2[extreme]
    
    <span class="co"># Matrix form of covariates</span>
    xg =<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">cbind</span>(xg1,xg2))
    xe =<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">cbind</span>(xe1,xe2))</code></pre></div>
<p>First we consider testing in the EPS-only design. Here are a few key examples.</p>
<p>Example 1.1) Test <code>H0: bg = 0</code> against <code>H1: bg ≠ 0</code> for each genetic covariate at a time. This is what you should use for a GWAS type analysis. Each genetic variable is tested against the null model <code>y = a + be*xe + e</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 1.1) Perform two-sided test for one genetic covariate at a time</span>
    <span class="kw">epsonly.test</span>(y~xe1+xe2, <span class="dt">SNP=</span><span class="kw">cbind</span>(xg1,xg2), <span class="dt">cutoffs=</span><span class="kw">c</span>(l,u))$p.value</code></pre></div>
<pre><code>## initial  value 4527.011996 
## final  value 4478.235271 
## converged</code></pre>
<pre><code>##          p.value
## xg1 2.723088e-06
## xg2 1.371253e-21</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># or using matrix form</span>
    <span class="kw">epsonly.test</span>(y~xe, <span class="dt">SNP=</span>xg, <span class="dt">cutoffs=</span><span class="kw">c</span>(l,u))$p.value</code></pre></div>
<pre><code>## initial  value 4527.011996 
## final  value 4478.235271 
## converged</code></pre>
<pre><code>##          p.value
## xg1 2.723088e-06
## xg2 1.371253e-21</code></pre>
<p>Example 1.2) Test <code>H0: bg = 0</code> against <code>H1: bg ≠ 0</code> for all genetic covariates simultaneously. (Only when the number of SNPs is low)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 1.2) Perform two-sided test for all genetic covariates simultanously</span>
    <span class="kw">epsonly.test</span>(y~xe1+xe2, <span class="dt">SNP=</span>xg, <span class="dt">cutoffs=</span><span class="kw">c</span>(l,u), <span class="dt">onebyone =</span> <span class="ot">FALSE</span>)$p.value</code></pre></div>
<pre><code>## initial  value 4527.011996 
## final  value 4478.235271 
## converged</code></pre>
<pre><code>##           p.value
## [1,] 6.164459e-25</code></pre>
<p>Example 1.3) Test gene-environment interactions</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 1.3) Test gene-environment interaction</span>
    <span class="kw">epsonly.testGE</span>(y~xe1+xe2+xg1+xg2, <span class="dt">GE=</span><span class="kw">c</span>(<span class="st">&quot;xe1:xg1&quot;</span>), <span class="dt">cutoffs=</span><span class="kw">c</span>(l,u))$p.value</code></pre></div>
<pre><code>## initial  value 4468.107553 
## iter  10 value 4420.644126
## iter  10 value 4420.644124
## iter  10 value 4420.644124
## final  value 4420.644124 
## converged</code></pre>
<pre><code>##              p.value
## xe1:xg1 1.651655e-05</code></pre>
<p>Next we consider fitting a linear regression model of the form <code>y = a + be*xe + bg*xg + e</code> using data from the EPS-only design.</p>
<p>Example 1.4) Fit a linear regression model to obtain parameter estimates and 95% confidence intervals</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 1.4) Fit a linear model</span>
    <span class="kw">epsonly.lm</span>(y~xe1+xe2+xg1+xg2, <span class="dt">cutoffs=</span><span class="kw">c</span>(l,u))</code></pre></div>
<pre><code>## initial  value 4468.107553 
## iter  10 value 4420.644126
## iter  10 value 4420.644124
## iter  10 value 4420.644124
## final  value 4420.644124 
## converged</code></pre>
<pre><code>## $coefficients
## (intercept)         xe1         xe2         xg1         xg2 
##  50.1218889   4.9680468   7.9490986   0.2656154   0.5941143 
## 
## $ci
##             lower 95% ci upper 95% ci
## (intercept)   49.9117841   50.3319937
## xe1            4.8871891    5.0489045
## xe2            7.7368739    8.1613232
## xg1            0.1550631    0.3761677
## xg2            0.4752296    0.7129990
## 
## $sigma
##         
## 1.98708</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># alternatively with matrix inputs</span>
    <span class="kw">epsonly.lm</span>(y~xe+xg, <span class="dt">cutoffs=</span><span class="kw">c</span>(l,u))</code></pre></div>
<pre><code>## initial  value 4468.107553 
## iter  10 value 4420.644126
## iter  10 value 4420.644124
## iter  10 value 4420.644124
## final  value 4420.644124 
## converged</code></pre>
<pre><code>## $coefficients
## (intercept)       xexe1       xexe2       xgxg1       xgxg2 
##  50.1218889   4.9680468   7.9490986   0.2656154   0.5941143 
## 
## $ci
##             lower 95% ci upper 95% ci
## (intercept)   49.9117841   50.3319937
## xexe1          4.8871891    5.0489045
## xexe2          7.7368739    8.1613232
## xgxg1          0.1550631    0.3761677
## xgxg2          0.4752296    0.7129990
## 
## $sigma
##         
## 1.98708</code></pre>
<p>Example 1.5) Fit a linear regression model with an interaction term</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 1.5) Fit a linear model with an interaction term</span>
    <span class="kw">epsonly.lm</span>(y~xe1+xe2+xg1+xg2+xe2*xg1, <span class="dt">cutoffs=</span><span class="kw">c</span>(l,u))$coefficients</code></pre></div>
<pre><code>## initial  value 4467.659369 
## iter  10 value 4419.580237
## final  value 4419.580062 
## converged</code></pre>
<pre><code>## (intercept)         xe1         xe2         xg1         xg2     xe2:xg1 
##  50.1844949   4.9660072   7.7925567   0.2000675   0.5936670   0.1689104</code></pre>
</div>
<div id="eps-full" class="section level3">
<h3>EPS-full</h3>
<p>For the <strong>EPS-full design</strong>, the variables <code>y</code> and <code>xe</code> are observed for all individuals (in a population / random sample) while <code>xg</code> is observed only for the extreme phenotype individuals. The value of <code>xg</code> should be set to <code>NA</code> for non-extreme phenotype individuals for which the genotype is not observed. The functions for testing and model fitting are based on a likelihood model for missing at random (MAR) covariates when the missing-mechanism is ignorable for likelihood inference.</p>
<p>The following example will be used throughout for illustrating EPS-full functions.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># Generate example data set of the type EPS-full</span>

    N =<span class="st"> </span><span class="dv">5000</span> <span class="co"># Number of individuals in a population</span>
    xe1 =<span class="st"> </span><span class="kw">rnorm</span>(<span class="dt">n =</span> N, <span class="dt">mean =</span> <span class="dv">2</span>, <span class="dt">sd =</span> <span class="dv">1</span>) <span class="co"># Environmental covariate</span>
    xe2 =<span class="st"> </span><span class="kw">rbinom</span>(<span class="dt">n =</span> N, <span class="dt">size =</span> <span class="dv">1</span>, <span class="dt">prob =</span> <span class="fl">0.3</span>) <span class="co"># Environmental covariate</span>
    xg1 =<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>), N, <span class="kw">c</span>(<span class="fl">0.4</span>,<span class="fl">0.3</span>,<span class="fl">0.3</span>), <span class="dt">replace=</span><span class="ot">TRUE</span>) <span class="co"># SNP</span>
    xg2 =<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>), N, <span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">0.3</span>,<span class="fl">0.2</span>), <span class="dt">replace=</span><span class="ot">TRUE</span>) <span class="co"># SNP</span>
    
    <span class="co"># Model parameters</span>
    a =<span class="st"> </span><span class="dv">50</span>; be1 =<span class="st"> </span><span class="dv">5</span>; be2 =<span class="st"> </span><span class="dv">8</span>; bg1 =<span class="st"> </span><span class="fl">0.3</span>; bg2 =<span class="st"> </span><span class="fl">0.6</span>; sigma =<span class="st"> </span><span class="dv">2</span>
    
    <span class="co"># Generate response y</span>
    y =<span class="st"> </span><span class="kw">rnorm</span>(N, <span class="dt">mean =</span> a +<span class="st"> </span>be1*xe1 +<span class="st"> </span>be2*xe2 +<span class="st"> </span>bg1*xg1 +<span class="st"> </span>bg2*xg2, <span class="dt">sd =</span> sigma)
    
    <span class="co"># Identify extremes, here upper and lower 25% of population</span>
    u =<span class="st"> </span><span class="kw">quantile</span>(y,<span class="dt">probs =</span> <span class="dv">3</span>/<span class="dv">4</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
    l =<span class="st"> </span><span class="kw">quantile</span>(y,<span class="dt">probs =</span> <span class="dv">1</span>/<span class="dv">4</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
    extreme =<span class="st"> </span>(y &lt;<span class="st"> </span>l) |<span class="st"> </span>(y &gt;=<span class="st"> </span>u)
    
    <span class="co"># Create the EPS-full data set by setting</span>
    <span class="co"># the SNP values of non-extremes to NA</span>
    xg1[!extreme] =<span class="st"> </span><span class="ot">NA</span>
    xg2[!extreme] =<span class="st"> </span><span class="ot">NA</span>
    
    xg =<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">cbind</span>(xg1,xg2))
    xe =<span class="st"> </span><span class="kw">as.matrix</span>(<span class="kw">cbind</span>(xe1,xe2))</code></pre></div>
<p>Below, we present a few key examples of testing for the EPS-full design.</p>
<p>Example 2.1) Test <code>H0: bg = 0</code> against <code>H1: bg ≠ 0</code> for each genetic covariate at a time. This is what you should use for a GWA type analysis. Each genetic variable is considered for inclusion into the null model <code>y = a + be*xe + e</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 2.1) Perform score test for one genetic covariate at a time</span>
    <span class="kw">epsfull.test</span>(y~xe1+xe2, <span class="dt">SNP=</span><span class="kw">cbind</span>(xg1,xg2))$p.value</code></pre></div>
<pre><code>##          p.value
## xg1 6.141198e-09
## xg2 1.756928e-29</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># alternatively with matrix form</span>
    <span class="kw">epsfull.test</span>(y~xe, <span class="dt">SNP=</span>xg)$p.value</code></pre></div>
<pre><code>##          p.value
## xg1 6.141198e-09
## xg2 1.756928e-29</code></pre>
<p>Example 2.2) Test <code>H0: bg = 0</code> against <code>H1: bg ≠ 0</code> for all genetic covariates simultaneously. (Only when the number of SNPs is low)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 2.2) Perform score test for all genetic covariates simultanously</span>
    <span class="kw">epsfull.test</span>(y~xe1+xe2, <span class="dt">SNP=</span>xg,<span class="dt">onebyone =</span> <span class="ot">FALSE</span>)$p.value</code></pre></div>
<pre><code>##              [,1]
## [1,] 8.030732e-38</code></pre>
<p>Example 2.3) Test a gene-environment interaction</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 2.3) Test interactions</span>
    <span class="kw">epsfull.testGE</span>(y~xe1+xe2+xg1+xg2, <span class="dt">GE=</span><span class="kw">c</span>(<span class="st">&quot;xe1:xg1&quot;</span>))$p.value</code></pre></div>
<pre><code>## initial  value 17668.798467 
## iter  10 value 16395.636774
## final  value 15929.959210 
## converged
## initial  value 17664.467632 
## iter  10 value 16032.992734
## final  value 15928.932397 
## converged</code></pre>
<pre><code>##        xe1:xg1
## [1,] 0.1518442</code></pre>
<p>Next we consider fitting the linear regression model <code>y = a + be*xe + bg*xg + e</code> using data from the EPS-full design.</p>
<p>Example 2.4) Fit a linear regression model to obtain parameter estimates and 95% confidence intervals</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 2.4) Fit a linear model</span>
    <span class="kw">epsfull.lm</span>(y~xe1+xe2+xg1+xg2)</code></pre></div>
<pre><code>## Identified 2 SNP(s), 2 environmental covariate(s), 0 environmental interaction(s), and 0 gene-environment interaction(s).</code></pre>
<pre><code>## initial  value 17668.798467 
## iter  10 value 16395.636774
## final  value 15929.959210 
## converged</code></pre>
<pre><code>## $coefficients
## (intercept)         xe1         xe2         xg1         xg2 
##  50.0631611   4.9915135   8.0183500   0.2940287   0.5763867 
## 
## $ci
##             lower 95% ci upper 95% ci
## (intercept)   49.8920671   50.2342552
## xe1            4.9341074    5.0489197
## xe2            7.8952979    8.1414022
## xg1            0.2044213    0.3836360
## xg2            0.4807988    0.6719746
## 
## $sigma
##          
## 2.005685</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># alternatively with matrix inputs</span>
    <span class="co"># epsfull.lm(y~xe+xg)</span></code></pre></div>
<p>Example 2.5) Fit a linear regression model with gene-environment interactions</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># 2.5) Fit a linear model with an interaction term</span>
    <span class="kw">epsfull.lm</span>(y~xe1+xe2+xg1+xg2+xe1*xg2)$coefficients</code></pre></div>
<pre><code>## Identified 2 SNP(s), 2 environmental covariate(s), 0 environmental interaction(s), and 1 gene-environment interaction(s).</code></pre>
<pre><code>## initial  value 17679.989896 
## iter  10 value 16095.595954
## final  value 15929.467424 
## converged</code></pre>
<pre><code>## (intercept)         xe1         xe2         xg1         xg2     xe1:xg2 
## 50.12602704  4.96122835  8.01745302  0.29312443  0.49041262  0.04314955</code></pre>
<div id="eps-full-and-hardy-weinberg" class="section level4">
<h4>EPS-full and Hardy-Weinberg</h4>
<p>The EPS-full model is based on a likelihood model that incorporates the distribution of the covariate with missing values. In general, we assume that <code>xg</code> takes values 0, 1 or 2 with probabilities <code>p0</code>, <code>p1</code> and <code>p2</code>, where <code>p2 = 1 - p0 - p1</code>. The user can however specify that Hardy-Weinberg equilibrium is assumed, in that case <code>xg</code> takes values 0, 1 or 2 with probabilities <code>(1-q)^2</code>, <code>2q(1-q)</code> and <code>q^2</code> where <code>q</code> is the minor allele frequency (MAF). This frequency can be unknown or given by the user. If the frequency is unknown, the function will provide an estimate. The following example illustrates model fitting under the Hardy-Weinberg assumption.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># Generate example data set of the type EPS-full</span>
    <span class="co"># Assume Hardy-Weinberg</span>

    N =<span class="st"> </span><span class="dv">5000</span> <span class="co"># Number of individuals in a population</span>
    xe =<span class="st"> </span><span class="kw">rnorm</span>(<span class="dt">n =</span> N, <span class="dt">mean =</span> <span class="dv">2</span>, <span class="dt">sd =</span> <span class="dv">1</span>) <span class="co"># Environmental covariate</span>
    q =<span class="st"> </span><span class="fl">0.3</span> <span class="co"># minor allele frequency</span>
    xg =<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>),N,<span class="kw">c</span>((<span class="dv">1</span>-q)^<span class="dv">2</span>,<span class="dv">2</span>*q*(<span class="dv">1</span>-q),q^<span class="dv">2</span>), <span class="dt">replace =</span> <span class="ot">TRUE</span>) <span class="co"># SNP</span>
    
    <span class="co"># Model parameters</span>
    a =<span class="st"> </span><span class="dv">50</span>; be =<span class="st"> </span><span class="dv">5</span>; bg =<span class="st"> </span><span class="fl">0.5</span>; sigma =<span class="st"> </span><span class="dv">2</span>
    
    <span class="co"># Generate response y</span>
    y =<span class="st"> </span><span class="kw">rnorm</span>(N, <span class="dt">mean =</span> a +<span class="st"> </span>be*xe +<span class="st"> </span>bg*xg, <span class="dt">sd =</span> sigma)
    
    <span class="co"># Identify extremes, here upper and lower 25% of population</span>
    u =<span class="st"> </span><span class="kw">quantile</span>(y,<span class="dt">probs =</span> <span class="dv">3</span>/<span class="dv">4</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
    l =<span class="st"> </span><span class="kw">quantile</span>(y,<span class="dt">probs =</span> <span class="dv">1</span>/<span class="dv">4</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
    extreme =<span class="st"> </span>(y &lt;<span class="st"> </span>l) |<span class="st"> </span>(y &gt;=<span class="st"> </span>u)
    
    <span class="co"># Create the EPS-full data set by setting</span>
    <span class="co"># the SNP values of non-extremes to NA</span>
    xg[!extreme] =<span class="st"> </span><span class="ot">NA</span>
    
    <span class="co"># Fit a linear model with known minor allele frequency</span>
    <span class="kw">epsfull.lm</span>(y~xe+xg,<span class="dt">hwe =</span> <span class="ot">TRUE</span>, <span class="dt">maf =</span> q)</code></pre></div>
<pre><code>## Identified 1 SNP(s), 1 environmental covariate(s), 0 environmental interaction(s), and 0 gene-environment interaction(s).</code></pre>
<pre><code>## Hardy-Weinberg equilibrium assumed with known minor allele frequencies: 0.3</code></pre>
<pre><code>## initial  value 13075.227936 
## final  value 13004.750883 
## converged</code></pre>
<pre><code>## $coefficients
## (intercept)          xe          xg 
##  50.0486337   4.9912137   0.4446405 
## 
## $ci
##             lower 95% ci upper 95% ci
## (intercept)   49.9048946   50.1923727
## xe             4.9361251    5.0463022
## xg             0.3318888    0.5573923
## 
## $sigma
##          
## 2.025588</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># Fit a linear model with unknown minor allele frequency</span>
    <span class="kw">epsfull.lm</span>(y~xe+xg,<span class="dt">hwe =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## Identified 1 SNP(s), 1 environmental covariate(s), 0 environmental interaction(s), and 0 gene-environment interaction(s).</code></pre>
<pre><code>## Hardy-Weinberg equilibrium assumed, unknown minor allele frequency.</code></pre>
<pre><code>## initial  value 13239.661270 
## iter  10 value 13004.179963
## iter  10 value 13004.179948
## iter  10 value 13004.179947
## final  value 13004.179947 
## converged</code></pre>
<pre><code>## $coefficients
## (intercept)          xe          xg 
##  50.0450423   4.9913106   0.4452022 
## 
## $ci
##             lower 95% ci upper 95% ci
## (intercept)   49.9007714   50.1893131
## xe             4.9362246    5.0463967
## xg             0.3325471    0.5578572
## 
## $sigma
##          
## 2.025405 
## 
## $maf
##           
## 0.3069139</code></pre>
</div>
<div id="eps-full-and-confounding-effects" class="section level4">
<h4>EPS-full and confounding effects</h4>
<p>If there are confounding effects in the data, then this must be accounted for in the distribution of <code>xg</code> in the EPS-full model. For example, if the distribution of a SNP in the population is different between men and women, then the estimation of the distribution of the SNP genotype must be conditioned on the sex of each individual. If not, the SNP can become a mediator for the effect of gender on phenotype, which will result in bias and false positives.</p>
<p>The functions for EPS-only with confounding lets the user type in which environmental covariates <code>xe</code> that are confounders for SNPs. In this package, confounders can only be discrete environmental covariates.</p>
<p>The following data set is created such that one SNP is dependent upon a binary covariate. The tests for associations show that a spurious result is found when confounding is not accounted for.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># Generate example data set of the type EPS-full</span>
    <span class="co"># with confounding</span>

    N =<span class="st"> </span><span class="dv">5000</span> <span class="co"># Number of individuals in a population</span>
    xe1 =<span class="st"> </span><span class="kw">rnorm</span>(<span class="dt">n =</span> N, <span class="dt">mean =</span> <span class="dv">2</span>, <span class="dt">sd =</span> <span class="dv">1</span>) <span class="co"># Environmental covariate</span>
    xe2 =<span class="st"> </span><span class="kw">rbinom</span>(<span class="dt">n =</span> N, <span class="dt">size =</span> <span class="dv">1</span>, <span class="dt">prob =</span> <span class="fl">0.3</span>) <span class="co"># Environmental covariate</span>
    xg1 =<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>), N, <span class="kw">c</span>(<span class="fl">0.4</span>,<span class="fl">0.3</span>,<span class="fl">0.3</span>), <span class="dt">replace =</span> <span class="ot">TRUE</span>) <span class="co"># SNP</span>
    
    xg2 =<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>,N) <span class="co"># SNP</span>
    n0 =<span class="st"> </span><span class="kw">sum</span>(xe2 ==<span class="st"> </span><span class="dv">0</span>)
    n1 =<span class="st"> </span><span class="kw">sum</span>(xe2 ==<span class="st"> </span><span class="dv">1</span>)
    xg2[xe2 ==<span class="st"> </span><span class="dv">0</span>] =<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>),n0,<span class="kw">c</span>(<span class="fl">0.5</span>,<span class="fl">0.3</span>,<span class="fl">0.2</span>), <span class="dt">replace =</span> <span class="ot">TRUE</span>)
    xg2[xe2 ==<span class="st"> </span><span class="dv">1</span>] =<span class="st"> </span><span class="kw">sample</span>(<span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>),n1,<span class="kw">c</span>(<span class="fl">0.3</span>,<span class="fl">0.1</span>,<span class="fl">0.6</span>), <span class="dt">replace =</span> <span class="ot">TRUE</span>)
    
    <span class="co"># Model parameters</span>
    a =<span class="st"> </span><span class="dv">50</span>; be1 =<span class="st"> </span><span class="dv">5</span>; be2 =<span class="st"> </span><span class="dv">8</span>; bg1 =<span class="st"> </span><span class="fl">0.3</span>; sigma =<span class="st"> </span><span class="dv">2</span>
    
    <span class="co"># Generate response y (xg2 has no effect)</span>
    y =<span class="st"> </span><span class="kw">rnorm</span>(N, <span class="dt">mean =</span> a +<span class="st"> </span>be1*xe1 +<span class="st"> </span>be2*xe2 +<span class="st"> </span>bg1*xg1 , <span class="dt">sd =</span> sigma)
    
    <span class="co"># Identify extremes, here upper and lower 25% of population</span>
    u =<span class="st"> </span><span class="kw">quantile</span>(y,<span class="dt">probs =</span> <span class="dv">3</span>/<span class="dv">4</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
    l =<span class="st"> </span><span class="kw">quantile</span>(y,<span class="dt">probs =</span> <span class="dv">1</span>/<span class="dv">4</span>,<span class="dt">na.rm=</span><span class="ot">TRUE</span>)
    extreme =<span class="st"> </span>(y &lt;<span class="st"> </span>l) |<span class="st"> </span>(y &gt;=<span class="st"> </span>u)
    
    xg1[!extreme] =<span class="st"> </span><span class="ot">NA</span>
    xg2[!extreme] =<span class="st"> </span><span class="ot">NA</span>
    
    xg =<span class="st"> </span><span class="kw">cbind</span>(xg1,xg2)
    
    <span class="co"># Test (with confounding)</span>
    <span class="kw">epsfull.test</span>(y~xe1+xe2, <span class="dt">SNP=</span>xg, <span class="dt">confounder=</span><span class="ot">TRUE</span>, <span class="dt">cx=</span><span class="kw">c</span>(<span class="st">&quot;xe2&quot;</span>))</code></pre></div>
<pre><code>## Confounders: xe2</code></pre>
<pre><code>## $statistic
##             t
## xg1 53.504391
## xg2  1.069639
## 
## $parameter
##     d.o.f
## xg1     1
## xg2     1
## 
## $p.value
##          p.value
## xg1 2.580154e-13
## xg2 3.010270e-01</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    <span class="co"># Test (without confounding)</span>
    <span class="kw">epsfull.test</span>(y~xe1+xe2, <span class="dt">SNP=</span>xg)</code></pre></div>
<pre><code>## $statistic
##             t
## xg1 53.094589
## xg2  3.497465
## 
## $parameter
##     d.o.f
## xg1     1
## xg2     1
## 
## $p.value
##          p.value
## xg1 3.178668e-13
## xg2 6.146284e-02</code></pre>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
